### PR TITLE
DoNotMerge: Button: how expensive is it to use makeStyles instead of scss?

### PR DESCRIPTION
**Pull request by dzearing**
Want to evaluate a hybrid model of styling. This PR does this:

1. Uses makeStyles, which uses merge-styles under the hood.
2. Still uses flattened styles we introduced with the sass conversion, where the only potential input is theme. This is far better than how we use merge-styles today, which involves style props caching, which are easy to break.

Why this is good:

* SSR will work
* Child windows will work
* RTL flipping will work
* Specificity is normalized
* Opens up the door to compare/contrast different cssinjs renderers

Why this is less good:

* Still runtime evaluation of styles, though done only once per theme/window.

### CODE REVIEW

**Comment by dzearing** [comment]
eek builds!

File: packages/react-theme-provider/src/makeClasses.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * Modifier classnames are added to the root className when the state contains a truthy value
+ * of the same name. For example, when the primary flag is present, the "_primary" modifier
+ * class will be appended to the root className prop.
+ *
+ * Enum classnames are also added to the root className when teh state contains an enum value
+ * which matches the value in the className. for example, when the `size` enum value is `small`,
+ * the "_size_small" enum class will be appended to the root className prop.
+ */
+export const makeClasses = <TStyleSet extends { [key: string]: IStyle }>(
```
Comment (xugao) [inline packages/react-theme-provider/src/makeClasses.ts:35]: i think this belongs better in `react-compose`

File: packages/react-theme-provider/src/getTokens.ts [hunk 2]
```diff
@@ -63,12 +63,12 @@ export function getTokens(theme: Theme): Tokens {
...
 
       button: {
-        fontWeight: fonts?.medium?.fontWeight,
+        fontWeight: 600, // fonts?.medium?.fontWeight,
         fontSize: fonts?.medium?.fontSize,
         fontFamily: fonts?.medium?.fontFamily,
         iconSize: fonts?.mediumPlus?.fontSize,
         borderRadius: effects?.roundedCorner2,
-        focusColor: palette?.neutralSecondary,
+        focusColor: palette?.black,
```
Comment (xugao) [inline packages/react-theme-provider/src/getTokens.ts:71]: are you sure? in V7 we have are using `neutralSecondary` for the outline color. however the focus border style has changed in v8, so maybe this is correct

File: packages/react-theme-provider/src/makeClasses.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  styleOrFunction: TStyleSet | ((theme: ITheme) => TStyleSet),
+) => {
+  const useStyles = makeStyles(styleOrFunction);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return (state: any) => {
+    const classes = useStyles();
+    const classNames = Object.keys(classes);
```
Comment (xugao) [inline packages/react-theme-provider/src/makeClasses.ts:43]: i think these are slots, not classNames

File: packages/react-button/src/components/Button/Button.styles.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+      paddingBottom: 0, // 'var(--button-padding)',
+
+      transition: 'var(--button-transition)',
+      whiteSpace: 'var(--button-whiteSpace, nowrap)',
+
+      fontFamily: 'var(--button-fontFamily)',
+      fontSize: 'var(--button-fontSize)',
+
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      fontWeight: 'var(--button-fontWeight)' as any,
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.styles.ts:43]: ah yeah i am fixing this [here](https://github.com/microsoft/fluentui/pull/14696/files#diff-2cefdb4f50d42d742f727bda024ecde6). but since this PR is on hold, i should probably split it out

File: packages/react-button/src/components/Button/Button.styles.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      boxShadow: 'var(--button-boxShadow)',
+
+      width: 'var(--button-width)',
+      height: 'var(--button-height)',
+      minHeight: 'var(--button-minHeight)',
+
+      paddingLeft: 16, // 'var(--button-padding)',
+      paddingRight: 16, // 'var(--button-padding)',
+      paddingTop: 0, // 'var(--button-padding)',
+      paddingBottom: 0, // 'var(--button-padding)',
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.styles.ts:34]: why this change?

File: packages/react-button/src/components/Button/Button.styles.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    width: '100%',
+    maxWidth: '100%',
+  },
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  _size_smallest: {
+    '--button-minHeight': 'var(--button-size-smallest)',
+  },
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (xugao) [inline packages/react-button/src/components/Button/Button.styles.ts:155]: hm can the rule be updated to support underscore prefix?

File: packages/react-button/src/components/Button/Button.styles.ts (hunk 4 above)
Comment (xugao) [reply packages/react-button/src/components/Button/Button.styles.ts:43]: Here is the fix: https://github.com/microsoft/fluentui/pull/14773

File: packages/react-button/src/components/Button/Button.styles.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
+import { makeStyles } from '@fluentui/react-theme-provider';
+
+export const useButtonStyles = makeStyles({
+  displayName: 'Button',
```
[the remark this answers is not in the document]
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.styles.ts]: removed.

File: packages/react-theme-provider/src/makeClasses.ts (hunk 1 above)
Comment (dzearing) [reply packages/react-theme-provider/src/makeClasses.ts:35]: hmm I'm not convinced that's the right place. 

It depends on makeStyles, which depends on theme.

It relies on the renderer, provided by the themeprovider, to make the styles. (Separate PR.)

My thought is that makeClasses which is in react-compose should be removed. Let react-compose be about utilities that help with composition.

File: packages/react-theme-provider/src/getTokens.ts (hunk 2 above)
Comment (dzearing) [reply packages/react-theme-provider/src/getTokens.ts:71]: Pretty sure this is correct. `neutralSecondary` is not dark enough. I can confirm with design.

File: packages/react-button/src/components/Button/Button.styles.ts (hunk 4 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.styles.ts:42]: tyty

File: packages/react-button/src/components/Button/Button.styles.ts (hunk 5 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.styles.ts]: resolved as discussed.

File: packages/react-button/src/components/Button/Button.styles.ts (hunk 6 above)
Comment (dzearing) [reply packages/react-button/src/components/Button/Button.styles.ts:166]: possibly, or we disable it for the file.

**Comment by dzearing** [comment]
Will re-open once split button is updated.
